api/src/test/java/uk/gov/dhsc/htbhf/claimant/reporting/MIReporterTest.java

```diff
@@ -0,0 +1,126 @@
+    void shouldNotGetPostcodeDataOrUpdateClaimWhenPostcodeDataExists() {
+        PostcodeData postcodeData = aPostcodeDataObjectForPostcode(VALID_POSTCODE);
+        Claim claim = aValidClaimBuilder().postcodeData(postcodeData).build();
+
+        miReporter.reportClaim(claim);
+
+        verifyZeroInteractions(restTemplate);
+        verifyZeroInteractions(claimRepository);
+    }
+
+    @Test
+    void shouldSaveNotFoundPostcodeDataWhenPostcodeNotFound() {
+        Claim claim = aValidClaimBuilder().postcodeData(null).build();
+        given(restTemplate.getForObject(anyString(), eq(PostcodeDataResponse.class))).willThrow(new HttpClientErrorException(NOT_FOUND));
+
+        miReporter.reportClaim(claim);
+
+        assertThat(claim.getPostcodeData()).isEqualTo(PostcodeData.NOT_FOUND);
+        String postcode = claim.getClaimant().getAddress().getPostcode();
+        verify(restTemplate).getForObject(getExpectedPostcodeUrl(postcode), PostcodeDataResponse.class);
+        verify(claimRepository).save(claim);
+    }
+
+    @Test
+    void shouldRethrowErrorWhenUnableToCallPostcodesIo() {
+        Claim claim = aValidClaimBuilder().postcodeData(null).build();
+        String postcode = claim.getClaimant().getAddress().getPostcode();
+        HttpClientErrorException expectedException = new HttpClientErrorException(INTERNAL_SERVER_ERROR);
+        given(restTemplate.getForObject(anyString(), eq(PostcodeDataResponse.class))).willThrow(expectedException);
+
+        PostcodesIoClientException actualException = catchThrowableOfType(() -> miReporter.reportClaim(claim), PostcodesIoClientException.class);
+
+        assertThat(actualException).isNotNull();
+        assertThat(actualException.getCause()).isEqualTo(expectedException);
+        assertThat(claim.getPostcodeData()).isNull();
+
+        verify(restTemplate).getForObject(getExpectedPostcodeUrl(postcode), PostcodeDataResponse.class);
+        verifyZeroInteractions(claimRepository);
+    }
+
+    @Test
+    void shouldThrowExceptionIfPostcodeDataResponseIsNull() {
+        Claim claim = aValidClaimBuilder().postcodeData(null).build();
+        String postcode = claim.getClaimant().getAddress().getPostcode();
+        PostcodeDataResponse response = new PostcodeDataResponse(null);
+        given(restTemplate.getForObject(anyString(), eq(PostcodeDataResponse.class))).willReturn(response);
+
+        PostcodesIoClientException exception = catchThrowableOfType(() -> miReporter.reportClaim(claim), PostcodesIoClientException.class);
+
+        assertThat(exception).isNotNull();
+        assertThat(claim.getPostcodeData()).isNull();
+        verify(restTemplate).getForObject(getExpectedPostcodeUrl(postcode), PostcodeDataResponse.class);
+        verifyZeroInteractions(claimRepository);
+    }
+
+    private String getExpectedPostcodeUrl(String postcode) {
+        String postcodeWithoutSpace = postcode.replace(" ", "");
+        return POSTCODES_IO_BASE_URI + POSTCODES_IO_PATH + postcodeWithoutSpace;
```

Comment:
Won't this have two forward slashes in it, the first constant ends with one and the second starts with one?Is that right?

Reply:
It will in the test but the rest template doesn't care, I've updated the test value to not have a '/' at the end of the url.